### PR TITLE
Fix RBAC token and env for runtime barrier

### PR DESCRIPTION
**Pull request by abuccts**
Fix RBAC token and env for runtime barrier.

### CODE REVIEW

File: src/kube-runtime/src/init [hunk 1]
```diff
@@ -130,7 +130,12 @@ else
   # priority=0
   # We should use k8s client to communicate api server, need to change this when RBAC is enabled
   CHILD_PROCESS="FRAMEWORK_RETRIEVER"
-  curl -s $KUBE_APISERVER_ADDRESS/apis/frameworkcontroller.microsoft.com/v1/namespaces/default/frameworks/$FC_FRAMEWORK_NAME > framework.json
+  if [[ -z $KUBE_APISERVER_ADDRESS ]]; then
+    KUBE_APISERVER_ADDRESS="KUBERNETES_SERVICE_HOST:KUBERNETES_SERVICE_PORT"
+    AUTH_FLAG='-k -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)"'
```
Comment (Binyang2014) [inline src/kube-runtime/src/init]: Please test this command in container, seems not work.

**Comment by Binyang2014** [comment]
>           serviceAccountName: 'frameworkbarrier-account',

Can you change it to another name, such as runtime-account. Since it used by runtime, not only frameworkbarrier

---
Refers to: src/rest-server/src/models/v2/job/k8s.js:305 in ffdbbdb. [](commit_id = ffdbbdbf1566465450906d23d6f411fc521386a4, deletion_comment = False)

File: src/rest-server/src/models/v2/job/k8s.js [hunk 2]
```diff
@@ -315,7 +315,7 @@ const generateTaskRole = (taskRole, labels, config) => {
                 },
                 {
                   name: 'KUBE_APISERVER_ADDRESS',
-                  value: launcherConfig.apiServerUri,
+                  value: apiserver.token ? '' : launcherConfig.apiServerUri,
```
[the remark this answers is not in the document]
Comment (abuccts) [reply src/rest-server/src/models/v2/job/k8s.js]: the env may be wrong, token will be read always

File: src/kube-runtime/src/init [hunk 3]
```diff
@@ -130,7 +130,12 @@ else
   # priority=0
   # We should use k8s client to communicate api server, need to change this when RBAC is enabled
   CHILD_PROCESS="FRAMEWORK_RETRIEVER"
-  curl -s $KUBE_APISERVER_ADDRESS/apis/frameworkcontroller.microsoft.com/v1/namespaces/default/frameworks/$FC_FRAMEWORK_NAME > framework.json
+  if [[ -z $KUBE_APISERVER_ADDRESS ]]; then
+    KUBE_APISERVER_ADDRESS="KUBERNETES_SERVICE_HOST:KUBERNETES_SERVICE_PORT"
```
[the remark this answers is not in the document]
Comment (abuccts) [reply src/kube-runtime/src/init]: updated

File: src/kube-runtime/src/init (hunk 1 above)
Comment (abuccts) [reply src/kube-runtime/src/init]: updated

**Comment by abuccts** [comment]
> > ```
> >       serviceAccountName: 'frameworkbarrier-account',
> > ```
> 
> Can you change it to another name, such as runtime-account. Since it used by runtime, not only frameworkbarrier
> 
> Refers to: src/rest-server/src/models/v2/job/k8s.js:305 in [ffdbbdb](https://github.com/microsoft/pai/commit/ffdbbdbf1566465450906d23d6f411fc521386a4). [](commit_id = [ffdbbdb](https://github.com/microsoft/pai/commit/ffdbbdbf1566465450906d23d6f411fc521386a4), deletion_comment = False)

so you should remove the other parts in runtime when gang scheduling is disabled, the account is designed to be used by barrier

File: src/kube-runtime/src/init [hunk 4]
```diff
@@ -130,7 +130,12 @@ else
   # priority=0
   # We should use k8s client to communicate api server, need to change this when RBAC is enabled
   CHILD_PROCESS="FRAMEWORK_RETRIEVER"
-  curl -s $KUBE_APISERVER_ADDRESS/apis/frameworkcontroller.microsoft.com/v1/namespaces/default/frameworks/$FC_FRAMEWORK_NAME > framework.json
+  if [[ -z $KUBE_APISERVER_ADDRESS ]]; then
+    KUBE_APISERVER_ADDRESS="https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT"
+    set -- -k -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)"
+  fi
+  curl -s "$@" \
```
Comment (Binyang2014) [inline src/kube-runtime/src/init]: Please test this. runtime not use Bourne shell. It use busybox default shell

Comment (abuccts) [reply src/kube-runtime/src/init]: it's Bourne shell https://github.com/microsoft/pai/blob/master/src/kube-runtime/src/init#L1

File: src/rest-server/src/models/v2/job/k8s.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Binyang2014) [reply src/rest-server/src/models/v2/job/k8s.js]: Please add comments for this. It's not easy to understand why we need  to check token before pass apiServerUri

---
In reply to: [335337234](https://github.com/microsoft/pai/pull/3739#discussion_r335337234) [](ancestors = 335337234)

**Comment by Binyang2014** [comment]
We should check this later. Currently, just keep as it is.

---
In reply to: [542586626](https://github.com/microsoft/pai/pull/3739#issuecomment-542586626) [](ancestors = 542586626)

**Pull request merged**